File: app/GOExercise/GOExercise/Models/DailyStatus.swift
import Foundation

enum DailyStatus: String, Codable, CaseIterable, Sendable {
    case achieved
    case rest
    case missed
    case future
    case todayPending
    case todayAchieved
    /// フリーズ(保険チケット)で救済された日。達成扱いだが「実際に運動した日(◎)」とは
    /// 表示を分ける(ユーザー要望: ◎=実運動 / ○=フリーズ / 休=休養日)。
    case rescued

    var symbol: String {
        switch self {
        case .achieved: "◎"
        case .rescued: "○"
        case .rest: "休"
        case .missed: "×"
        case .future: "-"
        case .todayPending: "・"
        case .todayAchieved: "◎"
        }
    }

    var countsAsAchieved: Bool {
        self == .achieved || self == .rest || self == .todayAchieved || self == .rescued
    }
}

struct DailyStatusEntry: Identifiable, Hashable, Sendable {
    var id: Date { date }
    let date: Date
    let status: DailyStatus
    let recordIds: [UUID]
}

struct StreakState: Equatable, Sendable {
    let currentStreak: Int
    let longestStreak: Int
    let lastAchievedDate: Date?
}

struct WeeklyProgress: Equatable, Sendable {
    let achievedCount: Int
    let totalDays: Int

    var rate: Double {
        guard totalDays > 0 else { return 0 }
        return Double(achievedCount) / Double(totalDays)
    }
}

File: app-android/app/src/main/java/com/goexercise/app/domain/DailyStatus.kt
package com.goexercise.app.domain

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * 1 日の達成状態。iOS `DailyStatus`(String, Codable)の移植(symbol / countsAsAchieved を含む)。
 * @SerialName で iOS rawValue(achieved/rest/...)と JSON 表現を一致させる。
 */
@Serializable
enum class DailyStatus(val rawValue: String, val symbol: String) {
    // ◎=実運動 / ○=保険チケット救済 / 休=休養日(iOS と一致)。
    // rawValue は iOS DailyStatus.rawValue・@SerialName と同一文字列(クロスOS契約)。
    @SerialName("achieved") Achieved("achieved", "◎"),
    @SerialName("rescued") Rescued("rescued", "○"),
    @SerialName("rest") Rest("rest", "休"),
    @SerialName("missed") Missed("missed", "×"),
    @SerialName("future") Future("future", "-"),
    @SerialName("todayPending") TodayPending("todayPending", "・"),
    @SerialName("todayAchieved") TodayAchieved("todayAchieved", "◎");

    val countsAsAchieved: Boolean
        get() = this == Achieved || this == Rescued || this == Rest || this == TodayAchieved

    companion object {
        /**
         * rawValue 文字列から復元。未知の値(将来 iOS が追加した case 等)は Future にフォールバックして
         * 配列長(月→日7要素)を保つ。drop すると後続の曜日が左にズレるため(iOS 1.3 と同じ防御)。
         */
        fun fromRaw(raw: String): DailyStatus = entries.firstOrNull { it.rawValue == raw } ?: Future
    }
}

/** 1 日分の状態エントリ(週/月カレンダー用)。iOS `DailyStatusEntry` の移植。 */
data class DailyStatusEntry(
    val date: java.time.LocalDate,
    val status: DailyStatus,
    val recordIds: List<String>,
)

/** 連続記録の集計。iOS `StreakState` の移植。 */
data class StreakState(
    val currentStreak: Int,
    val longestStreak: Int,
    val lastAchievedDate: java.time.LocalDate?,
)

/** 週間達成率。iOS `WeeklyProgress` の移植。 */
data class WeeklyProgress(
    val achievedCount: Int,
    val totalDays: Int,
) {
    val rate: Double
        get() = if (totalDays > 0) achievedCount.toDouble() / totalDays.toDouble() else 0.0
}
